New cache file linking guide and related updates to existing docs

jorgeorpinel opened the pull request:
Partially addresses #265

jorgeorpinel commented:
@shcheklein the guide isn't here yet but you may want to double check the file changes in this PR so far (reflecting https://github.com/iterative/dvc/pull/1841), which are needed for consistency with recent updates I made (#344) and with the new user guides I'll push up here soon. Thanks ✌️

shcheklein commented on static/docs/get-started/add-files.md:
```diff
@@ -41,7 +41,7 @@ $ git commit -m "add source data to DVC"
 
 You can see that actual data file has been moved to the `.dvc/cache` directory
-(usually hardlink or reflink is created, so no physical copying is happening).
+(usually a reflink is created, so no file content copying is happening).
```
I would actually make an details section - something like: `IMPORTANT: Performance on large data or something` ...

jorgeorpinel replied:
OK. I added that tradeoff information to the new file link type guide (now pushed) and linked to it here instead.

shcheklein commented on static/docs/get-started/example-pipeline.md:
```diff
@@ -167,8 +167,8 @@ intermediate result.
 similar to Git's internal objects store but made specifically to handle large
-data files. DVC is using reflinks, hardlinks and other optimizations to manage
-your actual workspace without copying every time object from/to the cache.
+data files. DVC uses reflinks, hardlinks, or other optimizations to manage your
```
the same, we should be emphasizing this. It was really painful decision to disable hardlinks and symlinks. A lot of ppl might be confused and disappointed to see that we are copying their files. We should give a lot of warnings and docs so that they understand right away that there is a way to enable an optimization.

jorgeorpinel replied:
Split into special note.

shcheklein commented on static/docs/philosophy/related-technologies.md:
```diff
@@ -81,7 +81,8 @@ process.
      symlinks instead of the actual files.
 
-   - DVC uses hardlinks instead of symlinks to make user experience better.
+   - DVC uses reflinks (or hardlinks) instead of symlinks to make user
```
I think the previous paragraph should be updated. I don't see how two of them make sense together. @dmpetrov @efiop what did we want to actually say here? are you familiar with git-annex internals?

efiop replied:
@shcheklein Ah, right, this is an old one. Not sure why the first paragraph is that way. I would probably remove it for now.

Also, would rephrase a little: `DVC can use reflinks or hardlinks instead of symlinks to make user experience better.`

jorgeorpinel replied:
Updated.

shcheklein commented on static/docs/philosophy/related-technologies.md:
```diff
@@ -111,10 +112,9 @@ process.
-     `dvc checkout` command does not actually copy data files from cache to the
-     workspace, as copying files is a heavy operation for large files (30
-     GB+).
+   - DVC creates reflinks. The `dvc checkout` command does not actually copy
```
reflinks are rare. This phrase could be misleading. Btw, in places like this I would use a term copy-on-write links (aka reflinks)  and would put a link to some docs about them. It's not a very well known stuff.

jorgeorpinel replied:
Added as a foot note in this document but what would be a good source to link to about it? I haven't found anything that looks authoritative enough. Wikipedia? https://en.wikipedia.org/wiki/Copy-on-write

jorgeorpinel replied:
Link added.

jorgeorpinel commented:
OK guys I addressed the feedback but also had to push up all the other changes I have which add the new /doc/user-guide/cache-file-linking guide (See it [live in review app](https://dvc-org-pr-345.herokuapp.com/doc/user-guide/cache-file-linking)). Hopefully it's not too many changes to review now!

shcheklein commented on static/docs/user-guide/cache-file-linking.md:
```diff
@@ -0,0 +1,93 @@
+
+File links are entries in the file system that don't necessarily hold the file
+contents, but which point to where the file is actually stored. DVC uses file
+links in your project's workspace to avoid copying them from/to the DVC cache.
```
them - sounds like "to avoid copying links"

jorgeorpinel replied:
Changed to "avoid duplicating contents that will exist in the DVC cache."

shcheklein commented on static/docs/user-guide/cache-file-linking.md:
```diff
@@ -0,0 +1,93 @@
+`symlink` or soft link). While reflinks have all the benefits and none of the
+worries, they're not commonly supported in most platforms yet. Hard/soft links
+also optimize speed and space in the file system, but carry the risk of breaking
+your workflow ,since updating tracked files in the workspace causes data
```
` ,` -> `, `

shcheklein commented on static/docs/commands-reference/config.md:
```diff
@@ -103,58 +103,29 @@ files](/doc/user-guide/dvc-files-and-directories) for more details.)
-
-  4. **`copy`** - The most inefficient link type, yet the most widely supported
-     for any repo/cache FS combination. Suitable for scenarios with relatively
-     small data files, where copying them is not a performance/storage concern.
```
❤️

shcheklein commented on static/docs/get-started/example-pipeline.md:
```diff
@@ -167,8 +167,12 @@ intermediate result.
+data files.
+
+> **Note!** For performance with large data files, DVC can use file links from
+the cache to the workspace to avoid copying actual file contents. Refer to
```
did you run the pre-commit hook we have now? I think it puts `>` before each line

jorgeorpinel replied:
I had not but just installed the pre-commit hook (which uses `prettier`) and it does add them indeed. BTW Rewrap also puts `>` before each line but only if a long comment has no line breaks yet, otherwise it doesn't...

Also, `prettier` re-wraps my paragraphs differently than VSCode Rewrap. (The pre-commit hook doesn't line-break links.)

So there's inconsistencies throughout the docs with these 2 things at least. Maybe I should just run `prettier` on everything?

jorgeorpinel replied:
OK... I just ran ``npm bin`/prettier --write --prose-wrap=always` on all the files altered by this PR and the funny thing is that when `pre-commit` ran prettier again the results were different. So I'm not sure how to run `prettier` so it has the same settings as when it is ran by `pre-commit`.

Anyway, I noticed some problems with the auto-reformatting like that it changes how we wrote sidebar.json and it inserts more spaces to indent `yaml` code-blocks showing DVC-file examples so I'm turning it off for now. I'll ask about all this on Discord.

shcheklein commented on static/docs/get-started/example-versioning.md:
```diff
@@ -227,9 +227,10 @@ $ python train.py
-Note! `dvc remove` or `dvc unprotect` is required, otherwise `python train.py`
-will overwrite the existing file and may corrupt the cached version. Check this
-[guide](/doc/user-guide/update-tracked-file) to learn more.
+> **Note!** `dvc remove` or `dvc unprotect` is required, otherwise `python
```
it's not required anymore. It's required only in the protected mode. By default it's disabled and it's safe to modify/overwrite files.

jorgeorpinel replied:
Good point. Note removed.

shcheklein commented on static/docs/get-started/add-files.md:
```diff
@@ -41,7 +41,9 @@ $ git commit -m "add source data to DVC"
 
 You can see that actual data file has been moved to the `.dvc/cache` directory
-(usually hardlink or reflink is created, so no physical copying is happening).
+(ideally with reflinks if available on the system, otherwise by file copy – see
```
let's make it a separate `<details>` section - something like `Important performance caveats` or something like this.

jorgeorpinel replied:
Ok, done. This doc may have too many expandable detail sections now though, you may want to see it in the review app: https://dvc-org-pr-345.herokuapp.com/doc/get-started/add-files

shcheklein commented on static/docs/user-guide/update-tracked-file.md:
```diff
@@ -1,10 +1,14 @@
+Linking](/docs/user-guide/cache-file-linking)), updating tracked files has to be
+carried out with caution to avoid data corruption when the DVC config option
+`cache.type` is set to `hardlink` or/and `symlink`. (See `dvc config cache` for
+more details on setting the cache file link types.)
```
is it `).`?

jorgeorpinel replied:
In this case the entire sentence is inside the parenthesis so `.)` is correct 😃

I double checked just now and found several sources verifying this e.g. https://www.quickanddirtytips.com/education/grammar/periods-and-parentheses

```
If the words inside the parentheses are a complete sentence, the period, question
mark, or exclamation point that ends the sentence goes inside the parenthesis:

Bring chocolate. (Squiggly likes sweets.)
```

shcheklein commented on static/docs/user-guide/cache-file-linking.md:
```diff
@@ -0,0 +1,93 @@
+# File Link Types for the DVC Cache
```
I'm thinking if we should rename it to more user friendlier version - "Performance with large data" or something similar? Or are you thinking to put together a separate sections for that?

jorgeorpinel replied:
Yes! Actually this is not even the main new guide you requested in #265 **add a user guide page about the optimization (advanced mode)**. My plan was to add another guide for that matter, which will link back to this one.

Maybe I've added too much optimization info already to this one guide so no longer sure what will the optimization guide will contain. Is there anything else to optimize besides cache file link types?

shcheklein replied:
no, not that much .. I think this guide alone is probably enough. I would rename, may be write a brief motivation - that space and speed are important and we can't just copy, etc, etc. and call it a section :)

shcheklein commented on static/docs/commands-reference/config.md:
```diff
@@ -87,74 +87,45 @@ files](/doc/user-guide/dvc-files-and-directories) for more details.)
 - `cache.protected` - makes files in the workspace read-only. Possible values
   are `true` or `false` (default). Run `dvc checkout` for the change go into
-  effect. (It affects only files that are under DVC control.)
+  effect. (It affects only files that are under DVC control.)  
```
remove trailing whitespaces

jorgeorpinel replied:
They are in on purpose. In markdown they help to insert a line break within a paragraph.

Explained here: https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet#lists

shcheklein replied:
So, is this an actual whitespace? I'm fine with that but I would expect it to be fragile - some editors remove trailing whitespaces on save and prettier won't return them most likely. Not a big deal though, let's not worry about this now.

jorgeorpinel replied:
You're right. We may want to avoid these in the future... I'll search all the docs to see if there's many occurrences.

shcheklein commented on static/docs/commands-reference/config.md:
```diff
@@ -87,74 +87,45 @@ files](/doc/user-guide/dvc-files-and-directories) for more details.)
   the cached version of a file by editing or overwriting it. Turning this config
   option on forces you to run `dvc unprotect` before updating a file, providing
-  an additional layer of security to your data.
+  an additional layer of security to your data.  
```
trailing whitespaces

jorgeorpinel replied:
It's another line break. Should I remove these and add full paragraphs breaks instead?

shcheklein replied:
Ah, np then. I'm totally fine with two paragraph when you think it's appropriate.

jorgeorpinel commented:
Ready, @shcheklein

jorgeorpinel commented:
Ready, @shcheklein. Just a couple discussions still open above for a final decision on your part please. Thanks

shcheklein commented:
Great stuff, @jorgeorpinel ! 🎉
